lite/examples/video_classification/android/app/src/main/java/org/tensorflow/lite/examples/videoclassification/ml/VideoClassifier.kt

```diff
@@ -0,0 +1,255 @@
+
+        return inputs
+    }
+
+    /**
+     * Initialize the output objects to store the TFLite model outputs.
+     */
+    private fun initializeOutput(): HashMap<String, Any> {
+        val outputs = HashMap<String, Any>()
+        for (outputName in interpreter.getSignatureOutputs(SIGNATURE_KEY)) {
+            // Initialize a ByteBuffer to store the output of the TFLite model.
+            val tensor = interpreter.getOutputTensorFromSignature(outputName, SIGNATURE_KEY)
+            val byteBuffer = ByteBuffer.allocateDirect(tensor.numBytes())
+            byteBuffer.order(ByteOrder.nativeOrder())
+            outputs[outputName] = byteBuffer
+        }
+
+        return outputs
+    }
+
+    /**
+     * Run classify and return a list include action and score.
+     */
+    fun classify(inputBitmap: Bitmap): List<Category> {
+        // As this model is stateful, ensure there's only one inference going on at once.
+        synchronized(lock) {
+            // Prepare inputs.
+            val tensorImage = preprocessInputImage(inputBitmap)
+            inputState[IMAGE_INPUT_NAME] = tensorImage.buffer
+
+            // Initialize a placeholder to store the output objects.
+            val outputs = initializeOutput()
+
+            // Run inference using the TFLite model.
+            interpreter.runSignature(inputState, outputs)
+
+            // Post-process the outputs.
+            val categories = postprocessOutputLogits(outputs[LOGITS_OUTPUT_NAME] as ByteBuffer)
+
+            // Store the output states to feed as input for the next frame.
+            outputs.remove(LOGITS_OUTPUT_NAME)
+            inputState = outputs
+
+            // Sort the output and return only the top K results.
+            categories.sortByDescending { it.score }
+
+            // Take only maxResults number of result.
+            return categories.subList(0, max(maxResults, categories.size))
+        }
+    }
+
+    /**
+     * Return the input size required by the model.
+     */
+    fun getInputSize(): Size {
+        return Size(inputWidth, inputHeight)
+    }
+
+    /**
+     * Convert input bitmap to TensorImage and normalize.
+     */
+    private fun preprocessInputImage(bitmap: Bitmap): TensorImage {
+        val size = min(bitmap.width, bitmap.height)
+        val imageProcessor = ImageProcessor.Builder().apply {
+            add(ResizeWithCropOrPadOp(size, size))
+            add(ResizeOp(inputHeight, inputWidth, ResizeOp.ResizeMethod.BILINEAR))
+            add(NormalizeOp(INPUT_MEAN, INPUT_STD))
+        }.build()
+        val tensorImage = TensorImage(DataType.FLOAT32)
+        tensorImage.load(bitmap)
+        return imageProcessor.process(tensorImage)
+    }
+
+    /**
+     * Convert input bitmap to TensorImage and normalize.
+     */
+    private fun postprocessOutputLogits(logitsByteBuffer: ByteBuffer): MutableList<Category> {
+        // Convert ByteBuffer to FloatArray.
+        val tensorBuffer = TensorBuffer.createFixedSize(
```

Comment:
It seems to be overkilled to use TensorBuffer to convert the ByteBuffer into a float array. Does it work to project the ByteBuffer to FloatBuffer, and then get the float array? Alternatively, you can create a TensorBuffer to store the output, meaning passing the TensorBuffer to Interpreter instead of using ByteBuffer.

Reply:
Fixed by converting ByteBuffer directly to FloatArray. As the model outputs contain INT32 tensor, I can't use TensorBuffer for all outputs because INT32 isn't supported yet.